models/demos/tg/llama3_70b/tt/llama_attention_galaxy.py

```diff
@@ -467,3 +491,194 @@ def attn_selfout(
+
+        fused_query_key_value = ttnn.reshape(fused_query_key_value, (1, 1, seq_len, -1))
+
+        xs.deallocate(True)
+
+        (
+            query_layer,  # [bsz, n_local_heads, seq_len, head_dim]
+            key_layer,  # [bsz, n_local_kv_heads, seq_len, head_dim]
+            value_layer,  # [bsz, n_local_kv_heads, seq_len, head_dim]
+        ) = ttnn.experimental.nlp_create_qkv_heads(
+            fused_query_key_value,
+            num_heads=self.n_local_heads,
+            num_kv_heads=self.n_local_kv_heads,
+            transpose_k_heads=False,
+            memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        )
+
+        fused_query_key_value.deallocate(True)
+
+        # ROTARY EMBEDDINGS
+        # Q Rotary Embeddings
+        # query_layer: ttnn.Shape([1, 8, seq_len, 128]) -> [bsz, n_local_heads, seq_len, head_dim]
+        query_layer_ret = ttnn.experimental.rotary_embedding_llama(
+            query_layer, rot_mats[0], rot_mats[1], self.transformation_mats
+        )
+        query_layer.deallocate(True)
+
+        # K Rotary Embeddings
+        # key_layer: ttnn.Shape([1, 1, seq_len, 128]) -> [bsz, n_local_kv_heads, seq_len, head_dim]
+        key_layer_ret = ttnn.experimental.rotary_embedding_llama(
+            key_layer, rot_mats[0], rot_mats[1], self.transformation_mats
+        )
+        key_layer.deallocate(True)
+
+        # query_layer = ttnn.matmul(
+        #     query_layer,
+        #     rot_mats,
+        #     # program_config=self.model_config["ROT_MAT_MM_PROGCFG"],
+        #     memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        #     compute_kernel_config=self.COMPUTE_KERNEL_ROTARY,
+        # )
+
+        # key_layer = ttnn.matmul(
+        #     key_layer,
+        #     rot_mats,
+        #     memory_config=ttnn.DRAM_MEMORY_CONFIG,
+        #     compute_kernel_config=self.COMPUTE_KERNEL_ROTARY,
+        # )
+
+        return query_layer_ret, key_layer_ret, value_layer
+
+    def prefill_prepare_tensor_for_kv_cache(self, key_or_value_layer, user_id):
+        tensor_copy = ttnn.clone(key_or_value_layer)
+        # Get all tensors from multi-device tensor
+        tensors = ttnn.get_device_tensors(tensor_copy)
+        # Get only tensors from specific column chips
+        # Get every 4th tensor starting from user_id // 8
+        single_column_tensors = tensors[user_id // 8 :: 4]
+        # Create multi-device tensor
+        multi_device_tensor = ttnn.aggregate_as_tensor(single_column_tensors)
+
+        return multi_device_tensor
+
+    def prefill_attn_mqa(
+        self,
+        query_layer,
+        key_layer,
+        value_layer,
+        attn_masks,
+        user_id: int = 0,
+    ):
+        # FILL K CACHE
+        keys = self.layer_past[0]
+        # Fill cache expects batch in dim0
+        keys_reshaped = ttnn.reshape(keys, [self.batch_size_per_device_group, self.n_local_kv_heads, -1, self.head_dim])
+
+        single_user_key_layer = self.prefill_prepare_tensor_for_kv_cache(key_layer, user_id)
```

Comment:
I thought we wanted this method to extract both the per user/device group kvcache slice (on the 8 devices that belong to that group) and the corresponding key_layer slice (on the same 8 devices that belong to that group, the other columns have a replication of that key projection)?
So that we pass only the 8-device slice of the mesh (for both the keys_reshaped and the kay_layer) to the fill_cache op?

Reply:
Only creating mesh of key_layer and passing multi-device tensor(8x1 mesh) to fill_cache of keys_reshaped(8x4 mesh) works fine. It's tested so no need to create another multi-device tensor with 8x1 mesh. Just creates overhead.